x/pairing/keeper/msg_server_relay_payment.go

```diff
@@ -360,3 +362,61 @@ func appendRelayPaymentDetailsToEvent(from map[string]string, uniqueIdentifier u
 	}
 	return to
 }
+
+func (k Keeper) checkBadge(ctx sdk.Context, badgeData BadgeData, client string, relay *types.RelaySession) (newTimerExpiry uint64, err error) {
+	if !badgeData.Badge.IsBadgeValid(client, relay.LavaChainId, uint64(relay.Epoch)) {
+		return 0, utils.LavaFormatWarning("badge must match traits in relay request", fmt.Errorf("invalid badge"),
+			utils.Attribute{Key: "badgeAddress", Value: badgeData.Badge.Address},
+			utils.Attribute{Key: "badgeLavaChainId", Value: badgeData.Badge.LavaChainId},
+			utils.Attribute{Key: "badgeEpoch", Value: badgeData.Badge.Epoch},
+			utils.Attribute{Key: "relayAddress", Value: client},
+			utils.Attribute{Key: "relayLavaChainId", Value: relay.LavaChainId},
+			utils.Attribute{Key: "relayEpoch", Value: relay.Epoch},
+		)
+	}
+
+	badgeUsedCuKey := types.BadgeUsedCuKey(badgeData.Badge.ProjectSig, relay.Provider)
+	badgeUsedCuMapEntry, found := k.GetBadgeUsedCu(ctx, badgeUsedCuKey)
+	if !found {
+		// setting a timer with a callback to delete the badgeUsedCuEntry after badge.Epoch+blocksToSave (see keeper.go)
```

Comment:
is it setting a timer?
this is just a check, no stat change

Reply:
done. see https://github.com/lavanet/lava/commit/f1535db03c1f18bbf62b5f42b518426f41d7fbef